Make Mining Challenge component more modular

shaavan opened the pull request:
This PR is an effort in line to solve issue #448

According to [this](https://github.com/saving-satoshi/saving-satoshi/pull/527#pullrequestreview-1495974120) comment:

> By introducing all these kinds of variants we lose flexibility/composability. I was thinking we should do something more along these lines:
https://codesandbox.io/s/core-components-3tjy3n

This PR follows the example presented above to make the mining challenge component flexible and increase the compensability of components.

shaavan commented:
cc @afterburn

benalleng commented on content/lessons/chapter-2/mining-1.tsx:
```diff
@@ -187,53 +192,100 @@ export default function Mining1({ lang }) {
+                size={'small'}
+              />
+            </Card>
+            <Card className="w-max px-10 py-2.5">
```
same width issue here

shaavan replied:
Addressed the highlight in https://github.com/saving-satoshi/saving-satoshi/commit/f2792a38f471fb6a75da4b2dca348ada7cac9aab

A question related to the width issue that I have mentioned in this https://github.com/saving-satoshi/saving-satoshi/pull/535#issuecomment-1611343875.

benalleng commented on content/lessons/chapter-2/mining-1.tsx:
```diff
@@ -187,53 +192,100 @@ export default function Mining1({ lang }) {
+            </Card>
+            <Card className="w-max px-10 py-2.5">
+              <div
+                className={clsx('text-2xl font-normal', {
```
this needs to be `font-space-mono`

shaavan replied:
Addressed the highlight in https://github.com/saving-satoshi/saving-satoshi/commit/f2792a38f471fb6a75da4b2dca348ada7cac9aab

Thanks for catching it!

benalleng commented on content/lessons/chapter-2/mining-1.tsx:
```diff
@@ -187,53 +192,100 @@ export default function Mining1({ lang }) {
-                'fade-in text-white': blocks !== 0,
-                'text-black/50': blocks === 0,
+        <div className="fade-in grid w-full grid-cols-1 items-center px-[15px] py-[25px] md:order-last md:mx-[30px] md:my-0 md:w-[405px] md:p-[25px]">
+          <Card className="font-nunito text-lg font-semibold" transparent>
```
you will want to add a `mt-0` on this Card to cancel the top-margin for this first item

benalleng commented on ui/common/ProgressBar.tsx:
With us ultimately moving the progress down to 100 i don't think its necessary to do the full transform math instead lets use the simplified version 
```
style={{
    transform: `translate3d(-${total - progress}%,0,0)`,
}}
```

shaavan replied:
I agree with the change.

Addressed in b2fff5f90ad357681353f85f86442cf57c3d2d3b.
Thanks for the suggestion

benalleng reviewed (changes requested):
mostly styling changes but it all looks good! If we can get some approval of this PR i'll go ahead and close my PR with the other approach

shaavan commented:
Updated from e4faaa3216fd79d2731ed0a0068f1d9e8895cd1b to f2792a38f471fb6a75da4b2dca348ada7cac9aab ([pr535.01](https://github.com/shaavan/saving-satoshi/commits/pr535.01) -> [pr535.02](https://github.com/shaavan/saving-satoshi/commits/pr535.02), [diff](https://github.com/shaavan/saving-satoshi/compare/pr535.01..pr535.02))
Addressed @benalleng comments

**Updates:**
1. Added highlight attributes to Card as mentioned in the above [comments](https://github.com/saving-satoshi/saving-satoshi/pull/535#discussion_r1244165238).
2. Made appropriate font change.
3. Introduce the new `firstItem` prop to Card to set the appropriate margin for the first element.

@benalleng

> ..the width needs some adjustment as it is not ideal in mobile format

I tried noticing the difference between **Master Branch** and the **PR** on my iPhone 13, but except for a slight difference in the gap between the final two Cards, I couldn't observe any other change.

**Screenshot:**

|Master|PR|
|-------|---|
|![IMG_3905](https://github.com/saving-satoshi/saving-satoshi/assets/85434418/4d68c004-b0e4-4a4a-8858-bbc0d065a40c)|![IMG_3904](https://github.com/saving-satoshi/saving-satoshi/assets/85434418/12e25ac9-4670-4623-bcc0-86f4e2c81e9c)|

Are you pointing towards this difference in your comment, or is there some major difference noticeable on another device?

shaavan commented on content/lessons/chapter-2/mining-1.tsx:
```diff
@@ -187,53 +192,100 @@ export default function Mining1({ lang }) {
+            {/* progress = hashPower * 100/maxHashPower */}
+          </Card>
+          <div className="flex items-center justify-between gap-x-2.5 text-center font-space-mono">
+            <Card className="w-max px-3.5 py-2.5">
```
Addressed the highlight in f2792a38f471fb6a75da4b2dca348ada7cac9aab

A question related to the width issue that I have mentioned in this [comment](https://github.com/saving-satoshi/saving-satoshi/pull/535#issuecomment-1611343875).

shaavan commented:
Updated from f2792a38f471fb6a75da4b2dca348ada7cac9aab to 5e050bb126dbc4f3b1a3c043363874a35404b4c0 ([pr535.02](https://github.com/shaavan/saving-satoshi/commits/pr535.02) -> [pr535.03](https://github.com/shaavan/saving-satoshi/commits/pr535.03), [diff](https://github.com/shaavan/saving-satoshi/compare/pr535.02..pr535.03))
Addressed @benalleng comment

Update:

1. Fix the sizing issue of "dual" Cards.

Screenshot:

**pr535.03:**

![Screenshot 2023-06-29 at 4 50 28 PM](https://github.com/saving-satoshi/saving-satoshi/assets/85434418/f2b037b5-b3cf-4b17-b529-9f9df8ed8024)
![Screenshot 2023-06-29 at 4 50 40 PM](https://github.com/saving-satoshi/saving-satoshi/assets/85434418/780ab538-b649-410d-aff8-3ca89ba6136e)

benalleng reviewed (approved):
LGTM! I'm satisfied with the reusability of the components and believe that Card.tsx can be used to quickly complete the Profile component in #523
